src/platforms/common/server/kms_cpu_addressable_display_provider.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "kms_cpu_addressable_display_provider.h"
+#include "cpu_addressable_fb.h"
+#include <drm_fourcc.h>
+#include <xf86drm.h>
+
+namespace
+{
+auto drm_get_cap_checked(mir::Fd const& drm_fd, uint64_t cap) -> uint64_t
+{
+    uint64_t value;
+    if (drmGetCap(drm_fd, cap, &value))
+    {
+        BOOST_THROW_EXCEPTION((
+            std::system_error{
+                errno,
+                std::system_category(),
+                "Failed to query DRM capabilities"}));
+    }
+    return value;
+}
+}
+
+namespace mg = mir::graphics;
+namespace geom = mir::geometry;
+
+mg::kms::CPUAddressableDisplayProvider::CPUAddressableDisplayProvider(mir::Fd drm_fd)
+    : drm_fd{std::move(drm_fd)},
+      supports_modifiers{drm_get_cap_checked(this->drm_fd, DRM_CAP_ADDFB2_MODIFIERS) == 1}
+{
+    if (!drm_get_cap_checked(this->drm_fd, DRM_CAP_DUMB_BUFFER))
+    {
+        BOOST_THROW_EXCEPTION((std::runtime_error{"DRM device doesn't support CPU buffers"}));
+    }
```

Comment:
I think we are checking in the wrong place and reporting by the wrong mechanism.

If the "DRM device doesn't support CPU buffers", then surely we should be returning a `nullptr` from `maybe_create_interface()` not propagating an exception. And to achieve that, we should move the `DRM_CAP_DUMB_BUFFER` check out of the constructor into `maybe_create_interface()`.